### DISCUSSION TITLE
PUDL pandas style guide

### DISCUSSION THREAD

**Discussion by bendnorman** [catalyst-cooperative/pudl#3792] [Ideas]
Any interest in developing a Catalyst/PUDL pandas style guide? My return to data wrangling with the gridstatus work reminded me how many ways there are to do an operation in pandas. Some examples:

- built in operators or pandas methods: `+` vs `.add()`
- filtering data: `df.query("a = 1")` , `df[df.a == 1]`, `df[df.a.eq(1)]`
- accessing columns via strings or attributes: `df["my_column"]`, `df.my_column`
- probably many others I'm not thinking of

I think we all have different styles right now which sometimes makes it challenging to learn about a new section of data wrangling code.

If we wanted to go wild we could even create a pre-commit linter.

**Comment by cmgosnell** [comment]
hey ben! i think this is a good idea overall! I'm curious if there are any resources that we can just point to instead of making these decisions ourselves. On a quick search I found [this one](https://github.com/joshlk/pandas_style_guide) which is relatively minimal but I like it overall. (of your examples though, it does seem to only answer your `df["my_column"]` (good) vs `df.my_column` (bad apparently) question) (also the last suggestion in this guide i think we break all the time)

Also, we did a million years ago take a first crack at an attempt to develop some [code standards](https://docs.google.com/document/d/1hx9OC8WE4mov5wOrIqf_vYUrAuNgRkUAp-vWVZicn5E/edit). This does have a small pandas section but it was never really fleshed out.

**Reply by zaneselvans** [reply]
It's not very extensive, but there's also [the `pandas-vet` flake8 extension](https://pandas-vet.readthedocs.io/en/stable/index.html), most of which has been integrated [into Ruff](https://docs.astral.sh/ruff/rules/#pandas-vet-pd).

**Reply by bendnorman** [reply]
These seem like great starting points! Looks like we already use some of pandas-vet ruff rules. I'll look into how to extend the rules.
